man/send.man
.TH SEND %manext1% 2022-12-22 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
send \- send an nmh message
.SH SYNOPSIS
.HP 5
.na
.B send
.RB [ \-help ]
.RB [ \-version ]
.RB [ \-alias
.IR aliasfile ]
.RB [ \-draft ]
.RB [ \-draftfolder
.IR +folder ]
.RB [ \-draftmessage
.IR msg ]
.RB [ \-nodraftfolder ]
.RB [ \-filter
.IR filterfile ]
.RB [ \-nofilter ]
.RB [ \-format " | " \-noformat ]
.RB [ \-forward " | " \-noforward ]
.RB [ \-mime " | " \-nomime ]
.RB [ \-msgid " | " \-nomsgid ]
.RB [ \-messageid
.IR localname " | " random ]
.RB [ \-push " | " \-nopush ]
.RB [ \-verbose " | " \-noverbose ]
.RB [ \-watch " | " \-nowatch ]
.RB [ \-mts
.IR smtp " | " sendmail/smtp " | " sendmail/pipe ]
.RB [ \-sendmail
.IR program ]
.RB [ \-server
.IR servername ]
.RB [ \-port
.IR port-name/number ]
.RB [ \-sasl ]
.RB [ \-nosasl ]
.RB [ \-saslmech
.IR mechanism ]
.RB [ \-authservice
.IR service ]
.RB [ \-snoop ]
.RB [ \-user
.IR username ]
.RB [ \-tls ]
.RB [ \-initialtls ]
.RB [ \-notls ]
.RB [ \-certverify ]
.RB [ \-nocertverify ]
.RB [ \-width
.IR columns ]
.RB [ file
\&...]
.ad
.SH DESCRIPTION
.B send
will cause each of the specified files to be delivered
to each of the destinations in the \*(lqTo:\*(rq, \*(lqcc:\*(rq,
\*(lqBcc:\*(rq, \*(lqDcc:\*(rq, and \*(lqFcc:\*(rq fields of the message.  If
.B send
is re-distributing a message, as invoked from
.BR dist ,
then the
corresponding \*(lqResent\-xxx\*(rq fields are examined instead.
.PP
By default,
.B send
uses the program
.B post
to do the actual
delivery of the messages, although this can be changed by defining the
.I postproc
profile component.  Most of the features attributed to
.B send
are actually performed by
.BR post .
.PP
Before
.B send
gives the message to
.B post
for delivery, the message is processed by
.B mhbuild
to perform any necessary MIME encoding of the outgoing message.  This
can be changed by the
.I buildmimeproc
profile component.
.B mhbuild
is invoked with the
.B \-auto
switch, so
.B mhbuild
directives are not processed by default.  See
.IR mhbuild (1)
for more information.
.PP
.B mhbuild
will scan the message draft for a header named
.IR Attach .
The draft is converted to a MIME message if one or more matches are found.
This conversion occurs before all other processing.
.IR whatnow (1)
describes the user interface for managing MIME attachments via
this mechanism.
.PP
The first part of the MIME message is the draft body if that body contains
any non-blank characters.
The body of each
.I Attach
header field is interpreted as a file name, and each file named is included as a separate
part in the MIME message.
.PP
Determination of the content MIME type inserted into the Content-Type
header for each part depends on how the
.B nmh
installation was configured.  If a program, such as
.B file
with a
.B \-\-mime
or
.B \-i
option, was found that can specify the type of a file as a MIME type
string, then that will be used.  To determine if your
.B nmh
was so configured, run
.B mhparam mimetypeproc
and see if a non-empty string is displayed.
.PP
If your
.B nmh
was not configured with a program to specify a file type as a MIME
string, then a different method is used to determine the content-type
string.  This method is also used if the configured mimetypeproc fails
to find the MIME type of the content.  For file names with dot suffixes,
the profile is scanned for a
.I mhshow-suffix-
entry for that suffix.
The content-type for the part is taken from that profile entry if a match is
found.  If a match is not found in the user profile, the mhn.defaults
profile is scanned next.
If no match is found or the file does not have a dot suffix, the content-type
is text/plain if the file contains only ASCII characters or application/octet-stream
if it contains characters outside of the ASCII range.  See
.IR mhshow (1)
for more details and example syntax.
.PP
Each attached MIME part contains a
\*(lqContent-Description\*(rq header that includes the filename, and
adds a \*(lqContent-Disposition\*(rq header.
Here is an example of MIME part headers for an attachment:
.PP
.nf
Content-Type: text/plain; name="VERSION"; charset="us-ascii"
Content-Description: VERSION
Content-Disposition: attachment; filename="VERSION"
.fi
.PP
See
.IR mhbuild (1)
for explanation of how the Content-Disposition value is selected.
.PP
If
.B \-push
is specified,
.B send
will detach itself from the user's
terminal and perform its actions in the background.  If
.BR push 'd
and the draft can't be sent, then an error message will be sent (using
the mailproc) back to the user.  If
.B \-forward
is given, then a copy
of the draft will be attached to this failure notice.  Using
.B \-push
differs from putting
.B send
in the background because the output is
trapped and analyzed by
.BR nmh .
.PP
If
.B \-verbose
is specified,
.B send
will indicate the interactions
occurring with the transport system, prior to actual delivery.
If
.B \-watch
is specified
.B send
will monitor the delivery of local
and network mail.  Hence, by specifying both switches, a large detail
of information can be gathered about each step of the message's entry
into the transport system.
.PP
The
.B \-draftfolder
.I +folder
and
.B \-draftmessage
.I msg
switches invoke
the
.B nmh
draft folder facility.  This is an advanced (and highly
useful) feature.  Consult
.IR mh-draft (5)
for more
information.
.PP
.B send
with no
.I file
argument will query whether the draft
is the intended file, whereas
.B \-draft
will suppress this question.
Once the transport system has successfully accepted custody of the
message, the file will be renamed with a site-dependent prefix
(usually a comma), which allows
it to be retrieved until the next draft message is sent.  If there are
errors in the formatting of the message,
.B send
will abort with a
(hopefully) helpful error message.
.PP
If a \*(lqBcc:\*(rq field is encountered, its addresses will be used for
delivery, and the \*(lqBcc:\*(rq field will be removed from the message
sent to sighted recipients.  The blind recipients will receive an entirely
new message with a minimal set of headers. The body of this new message
will contain a copy of the message sent to the sighted recipients, either
marked up with the indicator text "Blind-Carbon-Copy" or encapsulated
as a MIME digest.
.PP
If a \*(lqDcc:\*(rq field is encountered and the
.B sendmail/pipe
mail transport method is not in use, its addresses will be used for
delivery, and the \*(lqDcc:\*(rq field will be removed from the message.  The
blind recipients will receive exactly the same message as the sighted
recipients.  *WARNING* Recipients listed in the \*(lqDcc:\*(rq field receive no
explicit indication that they have received a \*(lqblind copy\*(rq.
This can cause blind recipients to
inadvertently reply to all of the sighted recipients of the
original message, revealing that they received a blind copy.
On the other hand, since a normal reply to a message sent
via a \*(lqBcc:\*(rq field
will generate a reply only to the sender of the original message,
it takes extra effort in most mailers to reply to the included
message, and so would usually only be done deliberately, rather
than by accident.
.PP
If the
.B sendmail/pipe
mail transport method is used, then messages containing
a \*(lqDcc:\*(rq field are rejected.

.PP
If
.B \-filter
.I filterfile
is specified, then this copy is filtered
(re-formatted) by
.B mhl
prior to being sent to the blind recipients.
Alternately, if you specify the
.B \-mime
switch, then
.B send
will
use the MIME rules for encapsulation.
.PP
Prior to sending the message, the \*(lqDate:\ now\*(rq field will be appended to the headers in the message.
If
.B \-msgid
is specified, then a \*(lqMessage\-ID:\*(rq field will also
be added to the message.
.PP
The
.B \-messageid
switch selects the style used for the part appearing after the @
in \*(lqMessage\-ID:\*(rq, \*(lqResent\-Message\-ID:\*(rq, and
\*(lqContent\-ID:\*(rq header fields.  The two acceptable options are
.B localname
(which is the default),
and
.BR random .
With
.BR localname ,
the local hostname is used.  With
.BR random ,
a random sequence of characters is used instead.  Note that the
.B \-msgid
switch must be enabled for this switch to have any effect.
.PP
If
.B send
is re-distributing a message (when invoked by
.BR dist ),
then \*(lqResent\-\*(rq will be prepended to each of these
fields: \*(lqFrom:\*(rq, \*(lqDate:\*(rq, and \*(lqMessage\-ID:\*(rq.
.PP
A \*(lqFrom:\*(rq field is required for all outgoing messages.  Multiple
addresses are permitted in the \*(lqFrom:\*(rq field, but a \*(lqSender:\*(rq
field is required in this case.  Otherwise a \*(lqSender:\*(rq field
is optional.
.PP
If a message with multiple \*(lqFrom:\*(rq
addresses does
.I not
include a \*(lqSender:\*(rq field but does include an \*(lqEnvelope\-From:\*(rq
field, the \*(lqEnvelope\-From:\*(rq field will be used to construct
a \*(lqSender:\*(rq field.
.PP
When using SMTP for mail submission, the envelope-from used for the SMTP
transaction is derived from the \*(lqEnvelope\-From:\*(rq field.
If no \*(lqEnvelope\-From:\*(rq field is present, the \*(lqSender:\*(rq
field is used.  If neither the \*(lqEnvelope\-From:\*(rq nor the
\*(lqSender:\*(rq field is present, the \*(lqFrom:\*(rq field is used.
When \*(lqEnvelope\-From:\*(rq appears in a message
it will be removed from the final outgoing message.
.PP
By using the
.B \-format
switch, each of the entries in the \*(lqTo:\*(rq
and \*(lqcc:\*(rq fields will be replaced with \*(lqstandard\*(rq
format entries.  This standard format is designed to be usable by all
of the message handlers on the various systems around the Internet.
If
.B \-noformat
is given, then headers are output exactly as they appear
in the message draft.
.PP
If an \*(lqFcc:\ folder\*(rq is encountered, the message will be copied
to the specified folder for the sender in the format in which it will
appear to any non\-Bcc receivers of the message.  That is, it will have
the appended fields and field reformatting.  The \*(lqFcc:\*(rq fields
will be removed from all outgoing copies of the message.
.PP
Beware that if an \*(lqFcc:\*(rq with one or more folders is present
but none of the folders exist, and the default
.I fileproc
and
.I postproc
are in use, then
.B refile
will prompt the user to create the folder(s) if
.B \-push
is not specified.  If all responses are negative, or creation of each folder
fails, or
.B \-push
is specified, the message will not be copied to any folder and will be
removed by
.BR post .
With the default
.B refile
switches, the message draft will be renamed according to the specification
of its
.B \-nolink
switch.
.PP
By using the
.B \-width
.I columns
switch, the user can direct
.B send
as to how long it should make header lines containing addresses.
.PP
The mail transport system default is provided in
.I %nmhetcdir%/mts.conf
but can be overridden here with the
.B \-mts
switch.
.PP
If nmh is using
.B sendmail/pipe
or
.B sendmail/smtp
as its mail transport system, the
.B \-sendmail
switch can be used to override the default
.B sendmail
program.
.PP
If nmh is using the SMTP MTA, the
.B \-server
and the
.B \-port
switches can be used to override the default mail server (defined by the
.I %nmhetcdir%/mts.conf
.I servers
entry).  The
.B \-snoop
switch can be used to view the SMTP transaction.  (Beware that the
SMTP transaction may contain authentication information either in
plaintext or easily decoded base64.)  If
.B \-sasl \-saslmech xoauth2
is used, the HTTP transaction is also shown.
.PP
If
.B nmh
has been compiled with SASL support, the
.B \-sasl
and
.B \-nosasl
switches will enable and disable
the use of SASL authentication with the SMTP MTA.  Depending on the
SASL mechanism used, this may require an additional password prompt from the
user (but the
.I netrc
file can be used to store this password, as described in
.IR mh-profile (5).
The
.B \-saslmech
switch can be used to select a particular SASL mechanism,
and the
.B \-user
switch can be used to select a authorization userid to provide to SASL
other than the default.  The credentials profile entry in
.IR mh\-profile (5)
describes the ways to supply a username and
password.
.PP
If SASL authentication is successful,
nmh
will attempt to negotiate a security layer for session encryption.
Encrypted data is labelled with `(encrypted)' and `(decrypted)' when
viewing the SMTP transaction with the
.B \-snoop
switch; see
.IR post (8)'s
description of
.B \-snoop
for its other features.
.PP
If
.B nmh
has been compiled with OAuth support, the
.B \-sasl
and
.B \-saslmech xoauth2
switches will enable OAuth authentication.  The
.B \-user
switch must be used, and the
.I username
must be an email address the user has for the service, which must
be specified with the
.B \-authservice
.I service
switch.  Before using OAuth authentication, the user must authorize nmh by
running
.B mhlogin
and grant authorization to that account.  See
.IR mhlogin (1)
for more details.
.PP
If
.B nmh
has been compiled with TLS support, the
.B \-tls
and
.B \-initialtls
switches will require the negotiation of TLS when
connecting to the SMTP MTA.  The
.B \-tls
switch will negotiate TLS as part of the normal SMTP protocol
using the STARTTLS command.  The
.B \-initialtls
will negotiate TLS immediately after the connection has
taken place, before any SMTP commands are sent or received.  Encrypted data
is labelled with `(tls-encrypted)' and
`(tls-decrypted)' when viewing the SMTP transaction with the
.B \-snoop
switch; see
.IR post (8)'s
description of
.B \-snoop
for its other features.
The
.B \-notls
switch will disable all attempts to negotiate TLS.
.PP
If port 465 is specified and none of the TLS switches were enabled,
.B \-initialtls
will be implied if TLS support was compiled in.  Though port 465 for
SMTPS (SMTP over SSL) was deregistered by IANA in 1998, it is still
used for that service.
.PP
When using TLS the default is to verify the remote certificate and SubjectName
against the local trusted certificate store.  This can be controlled by
the
.B \-certverify
and
.B \-nocertverify
switches.  See your OpenSSL documentation for more information on certificate
verification.
.PP
The files specified by the profile entry \*(lqAliasfile:\*(rq and any
additional alias files given by the
.B \-alias
.I aliasfile
switch will be

NEWS
#
# NEWS - Release notes for nmh 1.8
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been nearly five years since the last release of nmh, and there have
been a number of significant changes since that last release.  Long-time
MH and nmh users should read carefully the NOTABLE CHANGES section, as
there are some significant changes to nmh behavior.  Otherwise, please
see the README and INSTALL files for help on getting started with nmh.

This release is dedicated to Norman Z. Shapiro, co-designer of the MH
Message Handling System.  MH is the predecessor of nmh.  Norm was an
active supporter of nmh development until he passed away in October of
2021.  We are most grateful to Norm for his stewardship of MH and nmh.
    https://en.wikipedia.org/wiki/Norman_Shapiro

For news of future releases, subscribe to the low-volume
    https://lists.nongnu.org/mailman/listinfo/nmh-announce

---------------
NOTABLE CHANGES
---------------

- Support for Content-MD5 header fields, MIME content cache functionality,
  and the message/partial MIME type have been removed.
- Gmail OAuth2/XOAUTH support for desktop applications has been effectively
  dropped, so nmh no longer supports it.  nmh support for Gmail API access
  is experimental, please post to nmh-workers@nongnu.org if you'd like to
  help with test and development.
- repl(1) -convertargs now allows editing of the composition draft between
  translation and any encoding of text content.  Because encoding can wrap
  long lines, the use of a paragraph formatter has been removed from
  mhn.defaults.

------------
NEW FEATURES
------------

- The default editor has been changed from 'vi' to 'prompter', to align with
  historical practice and reduce packaging dependencies on external programs.
- A new -checkbase64 switch has been added to mhfixmsg(1).
- inc(1)/msgchk(1) now support STARTTLS for the POP protocol.
- All TLS-supported protocols now will send the SNI (server name indicator)
  TLS extension.
- A new mh-format function %(ordinal) has been implemented to output the
  appropriate ordinal suffix for numbers.  Example: "%(num 22)%(ordinal)"
  will output "22nd".
- show and mhl now decode more addresses in header fields.
- Added warning from all programs that read the profile if the profile
  contains a post entry, which is ignored, but does not contain a
  postproc entry.  In other words, if you get this warning and want
  to suppress it, your options include:
  1) Remove your post profile entry.
  2) Make your post profile entry a comment by prepending it with the #:
     comment indicator.
  3) Add a postproc entry that points to the post that you use.  That can
     be viewed with "mhparam postproc".
- scan(1) -file argument can be a Maildir directory.
- Updated mhn.defaults to prefer mpv(1) over xv(1) and replace mpeg_play(1),
  and to use it for all video types, not just video/mpeg.  And prefer all
  other searched-for pdf viewers over acroread(1).
- Added mhshow-suffix-video.mp4 to mhn.defaults, for use by mhshow(1) and
  send(1).
- Removed support from mhn.defaults for application/x-ivs and text/richtext.
- Changed interpretation of argument to mhfixmsg(1) -decodeheaderfieldbodies
  switch to specify character set of the decoded field bodies.
- repl(1) -convertargs now allows editing of the composition draft between
  translation and any encoding of text content.
- install-mh(1) now enables the mh-draft(5) draft folder facility.

-----------------
OBSOLETE FEATURES
-----------------

- The generation and verification of a Content-MD5 field has been removed
  without deprecation.  The related -check and -nocheck options now error.
- The MIME content cache functionality has been mostly non-functional since
  the development on nmh, and consequently all of the content caching code
  and related switches (-cache/-rcache/-wcache) have been removed.
- Support for generating and reassembling message/partial messages has been
  removed; it seems that this has been broken since 1.5 and there is very
  little support across MUAs.
- Marked Gmail OAuth2/XOAUTH support as being unsupported.
- Support for the MHPDEBUG environment variable was removed.  It was
  deprecated in nmh 1.7.  The pick(1) -debug switch replaced it.
- The 'libdir' mhparam(1) component was removed.  It was deprecated in
  nmh 1.7, when it was replaced by a new 'libexecdir' component.

---------
BUG FIXES
---------

- Fixed bcc to work with sendmail/pipe, and better documented that dcc
  doesn't work with it [Bug 55700].
- An -attendee switch has been added to mhical(1), for use when more than one
  (or zero) attendees match a user's mailbox.
- Fixed inc(1) and %(me) function escape to not obey Local-Mailbox profile
  component.
- Fixed source charset in mhfixmsg textcharset verbose output.
- Fixed mhfixmsg charset determination of content added with -reformat.
- Fixed file descriptor leak in mhfixmsg when run on multiple input files.
- Fixed mhfixmsg(1) -decodeheaderfilebodies to support mixed encoded/undecoded.
- Fixed memory corruption in post(1) and inc(1) when using XOAUTH2,
  with 4 or more entries in the oauth-authservice file.
- Added alias expansion to From: address for use by sendfrom.
- Removed extra space added before header field bodies by dist(1) to $mhdraft.
- Fixed display of iCalendar object with multiple VEVENTS using a VTIMEZONE.
- Fixed allowable encodings with MIME message types to get closer to RFC 2046.
- Detect other files regardless of backup prefix [Bug #49476].
- Copy if hard link by refile(1) or send(1) fails with EACESS [Bug #56575].

----------------------------------------------------------------------------

#
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time

uip/send.c
/* send.c -- send a composed message
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include "h/mh.h"
#include "sendsbr.h"
#include "sbr/m_name.h"
#include "sbr/getarguments.h"
#include "sbr/read_switch_multiword.h"
#include "sbr/concat.h"
#include "sbr/seq_setprev.h"
#include "sbr/seq_save.h"
#include "sbr/showfile.h"
#include "sbr/smatch.h"
#include "sbr/cpydata.h"
#include "sbr/m_draft.h"
#include "sbr/m_convert.h"
#include "sbr/folder_read.h"
#include "sbr/context_save.h"
#include "sbr/context_find.h"
#include "sbr/brkstring.h"
#include "sbr/ambigsw.h"
#include "sbr/push.h"
#include "sbr/path.h"
#include "sbr/print_version.h"
#include "sbr/print_help.h"
#include "sbr/utils.h"
#include "sbr/arglist.h"
#include "sbr/error.h"
#include <fcntl.h>
#include "sbr/done.h"
#ifdef OAUTH_SUPPORT
#include "sbr/oauth.h"
#endif
#include "sbr/m_maildir.h"
#include "sbr/m_mktemp.h"
#include "sbr/globals.h"

#ifndef CYRUS_SASL
# define SASLminc(a) (a)
#else /* CYRUS_SASL */
# define SASLminc(a)  0
#endif /* CYRUS_SASL */

#ifndef TLS_SUPPORT
# define TLSminc(a)  (a)
#else /* TLS_SUPPORT */
# define TLSminc(a)   0
#endif /* TLS_SUPPORT */

#define SEND_SWITCHES \
    X("alias aliasfile", 0, ALIASW) \
    X("debug", -5, DEBUGSW) \
    X("draft", 0, DRAFTSW) \
    X("draftfolder +folder", 6, DFOLDSW) \
    X("draftmessage msg", 6, DMSGSW) \
    X("nodraftfolder", 0, NDFLDSW) \
    X("filter filterfile", 0, FILTSW) \
    X("nofilter", 0, NFILTSW) \
    X("format", 0, FRMTSW) \
    X("noformat", 0, NFRMTSW) \
    X("forward", 0, FORWSW) \
    X("noforward", 0, NFORWSW) \
    X("mime", 0, MIMESW) \
    X("nomime", 0, NMIMESW) \
    X("msgid", 0, MSGDSW) \
    X("nomsgid", 0, NMSGDSW) \
    X("push", 0, PUSHSW) \
    X("nopush", 0, NPUSHSW) \
    X("unique", -6, UNIQSW) \
    X("nounique", -8, NUNIQSW) \
    X("verbose", 0, VERBSW) \
    X("noverbose", 0, NVERBSW) \
    X("watch", 0, WATCSW) \
    X("nowatch", 0, NWATCSW) \
    X("width columns", 0, WIDTHSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \
    X("dashstuffing", -12, BITSTUFFSW) \
    X("nodashstuffing", -14, NBITSTUFFSW) \
    X("client host", -6, CLIESW) \
    X("server host", 6, SERVSW) \
    X("snoop", 5, SNOOPSW) \
    X("sasl", SASLminc(4), SASLSW) \
    X("nosasl", SASLminc(6), NOSASLSW) \
    X("saslmech mechanism", SASLminc(6), SASLMECHSW) \
    X("authservice", SASLminc(0), AUTHSERVICESW) \
    X("user username", SASLminc(-4), USERSW) \
    X("port server-port-name/number", 4, PORTSW) \
    X("tls", TLSminc(-3), TLSSW) \
    X("initialtls", TLSminc(-10), INITTLSSW) \
    X("notls", TLSminc(-5), NTLSSW) \
    X("certverify", TLSminc(-10), CERTVERSW) \
    X("nocertverify", TLSminc(-12), NOCERTVERSW) \
    X("sendmail program", 0, MTSSM) \
    X("mts smtp|sendmail/smtp|sendmail/pipe", 2, MTSSW) \
    X("messageid localname|random", 2, MESSAGEIDSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(SEND);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(SEND, switches);
#undef X

#define USE_SWITCHES \
    X("no", 0, NOSW) \
    X("yes", 0, YESW) \
    X("list", 0, LISTDSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(USE);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(USE, anyl);
#undef X

extern bool debugsw;		/* from sendsbr.c */


int
main (int argc, char **argv)
{
    int msgp = 0, vecp;
    int isdf = 0;
    int msgnum, status;
    char *cp, *dfolder = NULL, *maildir = NULL;
    char buf[BUFSIZ], **ap, **argp, **arguments, *program;
    char *msgs[MAXARGS], **vec;
    const char *user = NULL, *saslmech = NULL;
    struct msgs *mp;
    struct stat st;
    char *auth_svc = NULL;

    if (nmh_init(argv[0], true, true)) { return 1; }

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    vec = argsplit(postproc, &program, &vecp);

    vec[vecp++] = "-library";
    vec[vecp++] = mh_xstrdup(m_maildir(""));

    if ((cp = context_find ("fileproc"))) {
	vec[vecp++] = "-fileproc";
	vec[vecp++] = cp;
    }

    if ((cp = context_find ("mhlproc"))) {
	vec[vecp++] = "-mhlproc";
	vec[vecp++] = cp;
    }

    if ((cp = context_find ("credentials"))) {
	/* post doesn't read context so need to pass credentials. */
	vec[vecp++] = "-credentials";
	vec[vecp++] = cp;
    }

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    die("-%s unknown\n", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s [file] [switches]", invo_name);
		    print_help (buf, switches, 1);
		    done (0);
		case VERSIONSW:
		    print_version(invo_name);
		    done (0);

		case DRAFTSW: 
		    msgs[msgp++] = draft;
		    continue;

		case DFOLDSW: 
		    if (dfolder)
			die("only one draft folder at a time!");
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    dfolder = path (*cp == '+' || *cp == '@' ? cp + 1 : cp,
			    *cp != '@' ? TFOLDER : TSUBCWF);
		    continue;
		case DMSGSW: 
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    msgs[msgp++] = cp;
		    continue;
		case NDFLDSW: 
		    dfolder = NULL;
		    isdf = NOTOK;
		    continue;

		case PUSHSW: 
		    pushsw = true;
		    continue;
		case NPUSHSW: 
		    pushsw = false;
		    continue;

		case UNIQSW: 
		    unique = true;
		    continue;
		case NUNIQSW: 
		    unique = false;
		    continue;

		case FORWSW:
		    forwsw = true;
		    continue;
		case NFORWSW:
		    forwsw = false;
		    continue;

		case VERBSW: 
		    verbsw = true;
		    vec[vecp++] = --cp;
		    continue;
		case NVERBSW:
		    verbsw = false;
		    vec[vecp++] = --cp;
		    continue;

		case MIMESW:
		    vec[vecp++] = --cp;
		    continue;
		case NMIMESW:
		    vec[vecp++] = --cp;
		    continue;

		case SNOOPSW:
		    vec[vecp++] = --cp;
		    continue;

		case DEBUGSW: 
		    debugsw = true;
		    /* FALLTHRU */
		case NFILTSW: 
		case FRMTSW: 
		case NFRMTSW: 
		case BITSTUFFSW:
		case NBITSTUFFSW:
		case MSGDSW: 
		case NMSGDSW: 
		case WATCSW: 
		case NWATCSW: 
		case SASLSW:
		case NOSASLSW:
		case TLSSW:
		case INITTLSSW:
		case NTLSSW:
		case CERTVERSW:
		case NOCERTVERSW:
		    vec[vecp++] = --cp;
		    continue;

		case USERSW:
		    vec[vecp++] = --cp;
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    vec[vecp++] = cp;
                    user = cp;
		    continue;

		case AUTHSERVICESW:
#ifdef OAUTH_SUPPORT
		    if (!(auth_svc = *argp++) || *auth_svc == '-')
			die("missing argument to %s", argp[-2]);
#else
		    die("not built with OAuth support");
#endif
		    continue;

		case SASLMECHSW:
		    if (!(saslmech = *argp) || *saslmech == '-')
			die("missing argument to %s", argp[-1]);
		    /* FALLTHRU */

		case ALIASW: 
		case FILTSW: 
		case WIDTHSW: 
		case CLIESW: 
		case SERVSW: 
		case PORTSW:
		case MTSSM:
		case MTSSW:
		case MESSAGEIDSW:
		    vec[vecp++] = --cp;
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    vec[vecp++] = cp;
		    continue;
	    }
	} else {
	    msgs[msgp++] = cp;
	}
    }

    /*
     * check for "Aliasfile:" profile entry
     */
    if ((cp = context_find ("Aliasfile"))) {
	for (ap = brkstring(mh_xstrdup(cp), " ", "\n"); ap && *ap; ap++) {
	    vec[vecp++] = "-alias";
	    vec[vecp++] = *ap;
	}
    }

    if (dfolder == NULL) {
	if (msgp == 0) {
	    msgs[msgp++] = mh_xstrdup(m_draft(NULL, NULL, 1, &isdf));
	    if (stat (msgs[0], &st) == -1)
		adios (msgs[0], "unable to stat draft file");
	    cp = concat ("Use \"", msgs[0], "\"? ", NULL);
	    for (status = LISTDSW; status != YESW;) {
		if (!(argp = read_switch_multiword (cp, anyl)))
		    done (1);
		switch (status = smatch (*argp, anyl)) {
		    case NOSW: 
			done (0);
		    case YESW: 
			break;
		    case LISTDSW: 
			showfile (++argp, msgs[0]);
			break;
		    default:
			inform("say what?");
			break;
		}
	    }
	} else {
	    for (msgnum = 0; msgnum < msgp; msgnum++)
		msgs[msgnum] = mh_xstrdup(m_maildir(msgs[msgnum]));
	}
    } else {
	if (!context_find ("path"))
	    free (path ("./", TFOLDER));

	if (!msgp)
	    msgs[msgp++] = "cur";
	maildir = m_maildir (dfolder);

	if (chdir (maildir) == -1)
	    adios (maildir, "unable to change directory to");

	/* read folder and create message structure */
	if (!(mp = folder_read (dfolder, 1)))
	    die("unable to read folder %s", dfolder);

	/* check for empty folder */
	if (mp->nummsg == 0)
	    die("no messages in %s", dfolder);

	/* parse all the message ranges/sequences and set SELECTED */
	for (msgnum = 0; msgnum < msgp; msgnum++)
	    if (!m_convert (mp, msgs[msgnum]))
		done (1);
	seq_setprev (mp);	/* set the previous-sequence */

	for (msgp = 0, msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++) {
	    if (is_selected (mp, msgnum)) {
		msgs[msgp++] = mh_xstrdup(m_name (msgnum));
		unset_exists (mp, msgnum);
	    }
	}

	mp->msgflags |= SEQMOD;
	seq_save (mp);
    }

#ifdef WHATNOW
go_to_it:
#endif /* WHATNOW */

    if ((cp = getenv ("SIGNATURE")) == NULL || *cp == 0)
	if ((cp = context_find ("signature")) && *cp)
	    setenv("SIGNATURE", cp, 1);

    for (msgnum = 0; msgnum < msgp; msgnum++)
	if (stat (msgs[msgnum], &st) == -1)
	    adios (msgs[msgnum], "unable to stat draft file");

    if ((annotext = getenv ("mhannotate")) == NULL || *annotext == 0)
	annotext = NULL;
    if (annotext && ((cp = getenv ("mhinplace")) != NULL && *cp != 0))
	inplace = atoi (cp);
    if ((altmsg = getenv ("mhaltmsg")) == NULL || *altmsg == 0)
	altmsg = NULL;	/* used by dist interface - see below */

    if ((cp = getenv ("mhdist"))
	    && *cp
	    && atoi(cp)
	    && altmsg) {
	vec[vecp++] = "-dist";
	if ((cp = m_mktemp2(altmsg, invo_name, NULL, NULL)) == NULL) {
	    die("unable to create temporary file");
	}
	distfile = mh_xstrdup(cp);
	(void) m_unlink(distfile);
	if (link (altmsg, distfile) == -1) {
	    /*
	     * Cygwin with FAT32 filesystem produces EPERM.
	     * Android produces EACCES.
	     */
	    if (errno != EXDEV  &&  errno != EPERM  &&  errno != EACCES
#ifdef EISREMOTE
		    && errno != EISREMOTE
#endif /* EISREMOTE */
		)
		adios (distfile, "unable to link %s to", altmsg);
	    free (distfile);
	    if ((cp = m_mktemp2(NULL, invo_name, NULL, NULL)) == NULL) {
		die("unable to create temporary file in %s",
		      get_temp_dir());
	    }
	    distfile = mh_xstrdup(cp);
	    {
		int in, out;
		struct stat st;

		if ((in = open (altmsg, O_RDONLY)) == -1)
		    adios (altmsg, "unable to open");
		fstat(in, &st);
		if ((out = creat (distfile, (int) st.st_mode & 0777)) == -1)
		    adios (distfile, "unable to write");
		cpydata (in, out, altmsg, distfile);
		close (in);
		close (out);
	    }	
	}
    } else {
	distfile = NULL;
    }

#ifdef OAUTH_SUPPORT
    if (auth_svc == NULL) {
        if (saslmech  &&  ! strcasecmp(saslmech, "xoauth2")) {
            die("must specify -authservice with -saslmech xoauth2");
        }
    } else {
        if (user == NULL) {
            die("must specify -user with -saslmech xoauth2");
        }
    }
#else
    NMH_UNUSED(auth_svc);
    NMH_UNUSED(user);
    NMH_UNUSED(saslmech);
#endif /* OAUTH_SUPPORT */

    if (altmsg == NULL || stat (altmsg, &st) == -1) {
	st.st_mtime = 0;
	st.st_dev = 0;
	st.st_ino = 0;
    }
    if (pushsw)
	push ();

    status = 0;

    for (msgnum = 0; msgnum < msgp; msgnum++) {
        switch (sendsbr (vec, vecp, program, msgs[msgnum], &st, 1, auth_svc)) {
	    case DONE: 
		done (++status);
		/* FALLTHRU */
	    case NOTOK: 
		status++;
		/* FALLTHRU */
	    case OK:
		break;
	}
    }

    context_save ();	/* save the context file */
    done (status);
    return 1;
}
